### PR TITLE
enhancement to clean up the control plane

### PR DESCRIPTION
**Pull request by xauthulei**
Signed-off-by: Leilei Hu <leileihu@cn.ibm.com>

- Fix the issue #71 
- Fix the commit error:

```
clusteradm % git commit -m "enhancement to clean up the control plane"
Desired copyright header is: Copyright Contributors to the Open Cluster Management project
FILE: SECURITY.md:
	- Need add Community copyright header to file
make: *** [check-copyright] Error 1
```

### CODE REVIEW

**Comment by xauthulei** [comment]
@qiujian16 ,Could you review this, Thanks 
Following are my testing snapshots :
```
llhu@huleis-mbp clusteradm % bin/clusteradm clean 
Error: hub not be initialized
llhu@huleis-mbp clusteradm % bin/clusteradm init 
The multicluster hub control plane has been initialized successfully!

You can now register cluster(s) to the hub control plane. Log onto those cluster(s) and run the following command:

    bin/clusteradm join --hub-token eyJhbGciOiJSUzI1NiIsImtpZCI6IjN6LUpLY0o1VFUwcnBIcTVYNnhJZVp4dmRfNldhRGtIRVg4M2hzY2otUXcifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJvcGVuLWNsdXN0ZXItbWFuYWdlbWVudCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VjcmV0Lm5hbWUiOiJjbHVzdGVyLWJvb3RzdHJhcC10b2tlbi1tcG16cyIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50Lm5hbWUiOiJjbHVzdGVyLWJvb3RzdHJhcCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6IjYxYzRkN2IyLWU3ZjMtNGU4OS04NDI1LTYyZjI5ZmFmYzA3MyIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpvcGVuLWNsdXN0ZXItbWFuYWdlbWVudDpjbHVzdGVyLWJvb3RzdHJhcCJ9.ji65stNZbecMDNB5ywlvSEWgTGis4vK26zbAE88KtzsU9KqmsbcKtp2ggZMnvg8_4hng-hcxsUvesAJ5lrwi0mafxjxrGiRCiNNggpO1ru6Z4_paX53JwPrrEqcI8ATky8dgthq-qlCCH0mbqX-leJ92y3lefB0XWw86ythh8ha4wRNwLzVYa6-NvRX-o0xIeB0AW4saZ2T4gNIyXa3Id6DozCtE89l-SsQJYyvQ07LRBuh8U8kYk_GgSh-uJ2dFfHAsFk8EJ85eDnGdkDxeHSxwHHRY9hFRVxq-3uIKENJllE4nJ0smhjBFZUbyx-VQnPBh_lCC8ZYaELkT6vh6gg --hub-apiserver https://127.0.0.1:61742 --cluster-name <cluster_name>

Replace <cluster_name> with a cluster name of your choice. For example, cluster1.

llhu@huleis-mbp clusteradm % kubectl get ns       
NAME                          STATUS   AGE
default                       Active   175m
kube-node-lease               Active   175m
kube-public                   Active   175m
kube-system                   Active   175m
local-path-storage            Active   175m
open-cluster-management       Active   17s
open-cluster-management-hub   Active   12s
llhu@huleis-mbp clusteradm % kubectl get crd      
NAME                                                           CREATED AT
clustermanagementaddons.addon.open-cluster-management.io       2021-11-11T08:59:05Z
clustermanagers.operator.open-cluster-management.io            2021-11-11T08:59:02Z
managedclusteraddons.addon.open-cluster-management.io          2021-11-11T08:59:06Z
managedclusters.cluster.open-cluster-management.io             2021-11-11T08:59:05Z
managedclustersetbindings.cluster.open-cluster-management.io   2021-11-11T08:59:06Z
managedclustersets.cluster.open-cluster-management.io          2021-11-11T08:59:05Z
manifestworks.work.open-cluster-management.io                  2021-11-11T08:59:06Z
placementdecisions.cluster.open-cluster-management.io          2021-11-11T08:59:07Z
placements.cluster.open-cluster-management.io                  2021-11-11T08:59:06Z
```
Verify:
```
llhu@huleis-mbp clusteradm % bin/clusteradm  --help 
Usage:
  clusteradm [command]

Available Commands:
  accept      accept a list of clusters
  clean       clean the hub
  config      Modify kubeconfig files
  delete      delete a resource
  enable      enable a feature
  get         get a resource
  help        Help about any command
  init        init the hub
  install     install a feature
  join        join a hub
  options     Print the list of flags inherited by all commands
  plugin      Provides utilities for interacting with plugins.
  unjoin      unjoin from a hub
  version     get the versions of the different components

Flags:
      --add-dir-header                   If true, adds the file directory to the header of the log messages
      --alsologtostderr                  log to standard error as well as files
      --as string                        Username to impersonate for the operation
      --as-group stringArray             Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
      --cache-dir string                 Default cache directory (default "/Users/llhu/.kube/cache")
      --certificate-authority string     Path to a cert file for the certificate authority
      --client-certificate string        Path to a client certificate file for TLS
      --client-key string                Path to a client key file for TLS
      --cluster string                   The name of the kubeconfig cluster to use
      --context string                   The name of the kubeconfig context to use
      --dry-run                          If set the generated resources will be displayed but not applied
  -h, --help                             help for clusteradm
      --insecure-skip-tls-verify         If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure
      --kubeconfig string                Path to the kubeconfig file to use for CLI requests.
      --log-backtrace-at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
      --log-dir string                   If non-empty, write log files in this directory
      --log-file string                  If non-empty, use this log file
      --log-file-max-size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                      log to standard error instead of files (default true)
      --match-server-version             Require server version to match client version
  -n, --namespace string                 If present, the namespace scope for this CLI request
      --one-output                       If true, only write logs to their native severity level (vs also writing to each lower severity level)
      --password string                  Password for basic authentication to the API server
      --request-timeout string           The length of time to wait before giving up on a single server request. Non-zero values should contain a corresponding time unit (e.g. 1s, 2m, 3h). A value of zero means don't timeout requests. (default "0")
  -s, --server string                    The address and port of the Kubernetes API server
      --skip-headers                     If true, avoid header prefixes in the log messages
      --skip-log-headers                 If true, avoid headers when opening log files
      --stderrthreshold severity         logs at or above this threshold go to stderr (default 2)
      --tls-server-name string           Server name to use for server certificate validation. If it is not provided, the hostname used to contact the server is used
      --token string                     Bearer token for authentication to the API server
      --user string                      The name of the kubeconfig user to use
      --username string                  Username for basic authentication to the API server
  -v, --v Level                          number for the log level verbosity
      --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging

Use "clusteradm [command] --help" for more information about a command.
llhu@huleis-mbp clusteradm % bin/clusteradm  clean  
2021/11/11 16:59:43 Wait to deleted resource cluster-manager.
2021/11/11 17:00:47 Recource cluster-manager have been deleted firstly.
The multicluster hub control plane has been clean up successfully!
llhu@huleis-mbp clusteradm % kubectl get crd        
No resources found
llhu@huleis-mbp clusteradm % kubectl get ns 
NAME                 STATUS   AGE
default              Active   178m
kube-node-lease      Active   178m
kube-public          Active   178m
kube-system          Active   178m
local-path-storage   Active   178m
```

File: pkg/cmd/clean/exec.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+		}
+	}
+
+	//Clean other resources
+	kubeClient, apiExtensionsClient, _, err := helpers.GetClients(f)
+	if err != nil {
+		return err
+	}
+	kubeClient.AppsV1().Deployments("open-cluster-management").Delete(context.Background(), "cluster-manager", metav1.DeleteOptions{})
```
Comment (qiujian16) [inline pkg/cmd/clean/exec.go:88]: I think you should delete the ClusterManager resource, and when it is removed, Delete the operator related resource

File: pkg/cmd/clean/exec.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	f := o.ClusteradmFlags.KubectlFactory
+	config, err := f.ToRESTConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+	clusterManagerClient, err := clustermanagerclient.NewForConfig(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = clusterManagerClient.OperatorV1().ClusterManagers().Delete(context.Background(), "cluster-manager", metav1.DeleteOptions{})
```
Comment (qiujian16) [inline pkg/cmd/clean/exec.go:68]: nit, i would think you have a flag to set name, but default is cluster-manager.

File: pkg/cmd/clean/exec.go (hunk 1 above)
Comment (qiujian16) [reply pkg/cmd/clean/exec.go:88]: you would not need to remove resources in open-cluster-management-hub if cluster-manager is deleted. I think you need to wait/watch until cluster-manager is cleaned at first, then delete the resources in open-cluster-management ns and also the klusterlet crd.

**Comment by xauthulei** [comment]
@qiujian16 ,  I have added a new flag `name` for setting the cluster manager resource name. Following my local testing snapshots:
```
llhu@huleis-mbp clusteradm % bin/clusteradm clean --help
clean the hub

Usage:
  clusteradm clean [flags]

Examples:

# Clean up the resource from the init stage
bin/clusteradm clean


Flags:
  -h, --help                  help for clean
      --name string           The name of the cluster manager resource (default "cluster-manager")
      --output-file string    The generated resources will be copied in the specified file
      --use-bootstrap-token   If set then the boostrap token will used instead of a service account token

Global Flags:
      --add-dir-header                   If true, adds the file directory to the header of the log messages
      --alsologtostderr                  log to standard error as well as files
      --as string                        Username to impersonate for the operation
      --as-group stringArray             Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
      --cache-dir string                 Default cache directory (default "/Users/llhu/.kube/cache")
      --certificate-authority string     Path to a cert file for the certificate authority
      --client-certificate string        Path to a client certificate file for TLS
      --client-key string                Path to a client key file for TLS
      --cluster string                   The name of the kubeconfig cluster to use
      --context string                   The name of the kubeconfig context to use
      --dry-run                          If set the generated resources will be displayed but not applied
      --insecure-skip-tls-verify         If true, the server's certificate will not be checked for validity. This will make your HTTPS connections insecure
      --kubeconfig string                Path to the kubeconfig file to use for CLI requests.
      --log-backtrace-at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
      --log-dir string                   If non-empty, write log files in this directory
      --log-file string                  If non-empty, use this log file
      --log-file-max-size uint           Defines the maximum size a log file can grow to. Unit is megabytes. If the value is 0, the maximum file size is unlimited. (default 1800)
      --logtostderr                      log to standard error instead of files (default true)
      --match-server-version             Require server version to match client version
  -n, --namespace string                 If present, the namespace scope for this CLI request
      --one-output                       If true, only write logs to their native severity level (vs also writing to each lower severity level)
      --password string                  Password for basic authentication to the API server
      --request-timeout string           The length of time to wait before giving up on a single server request. Non-zero values should contain a corresponding time unit (e.g. 1s, 2m, 3h). A value of zero means don't timeout requests. (default "0")
  -s, --server string                    The address and port of the Kubernetes API server
      --skip-headers                     If true, avoid header prefixes in the log messages
      --skip-log-headers                 If true, avoid headers when opening log files
      --stderrthreshold severity         logs at or above this threshold go to stderr (default 2)
      --tls-server-name string           Server name to use for server certificate validation. If it is not provided, the hostname used to contact the server is used
      --token string                     Bearer token for authentication to the API server
      --user string                      The name of the kubeconfig user to use
      --username string                  Username for basic authentication to the API server
  -v, --v Level                          number for the log level verbosity
      --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging
```
Verify :
```
llhu@huleis-mbp clusteradm % bin/clusteradm init        
The multicluster hub control plane has been initialized successfully!

You can now register cluster(s) to the hub control plane. Log onto those cluster(s) and run the following command:

    bin/clusteradm join --hub-token eyJhbGciOiJSUzI1NiIsImtpZCI6IlRhTVB0MlVDV2NxamdmT2pBS0lLM0tXWWlPak1zN2NpMlRua1NkNGFuQVkifQ.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJvcGVuLWNsdXN0ZXItbWFuYWdlbWVudCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VjcmV0Lm5hbWUiOiJjbHVzdGVyLWJvb3RzdHJhcC10b2tlbi1kbnptbCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50Lm5hbWUiOiJjbHVzdGVyLWJvb3RzdHJhcCIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6Ijk5YTA4NDM2LTAwNDAtNDkyMi05N2JjLWJkNDIyNTAxNjA0OSIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpvcGVuLWNsdXN0ZXItbWFuYWdlbWVudDpjbHVzdGVyLWJvb3RzdHJhcCJ9.YDyqQJE42G5x9Zlra5hv9IYG8_VRYxzo8TzufniWhXf7CipJKwgrvQhF7hs27WF1C__AX7ee5him90GQirutKlpyL4NKJGKFR9z7sjRUsE6fPi1ADfEeazvarC-NZVBl1nPCDRxgNGgw1zcoJvCOz0zICMIjT3uzURVLw50V9piE_pThNI8YtcZnvV9e_Z1-OCD8gMK5pq7r9Hcrvgajr2ZVnPKcJeilUkDRlh4InW4RUrcwDq3QN5tDG_OXTBeIdgmQcU-KfbCuV7nO9Q39jFX7dX3VCiOqnRtCHGFq6qKbco-oJFgfqjw96sPAOHzuXQTzz7kDkdgNdSrL0WSSVQ --hub-apiserver https://127.0.0.1:49606 --cluster-name <cluster_name>

Replace <cluster_name> with a cluster name of your choice. For example, cluster1.

llhu@huleis-mbp clusteradm % kubectl get ns
NAME                          STATUS   AGE
default                       Active   9m50s
kube-node-lease               Active   9m52s
kube-public                   Active   9m52s
kube-system                   Active   9m52s
local-path-storage            Active   9m41s
open-cluster-management       Active   16s
open-cluster-management-hub   Active   10s
llhu@huleis-mbp clusteradm % kubectl get crd
NAME                                                           CREATED AT
clustermanagementaddons.addon.open-cluster-management.io       2021-11-18T06:23:17Z
clustermanagers.operator.open-cluster-management.io            2021-11-18T06:23:13Z
managedclusteraddons.addon.open-cluster-management.io          2021-11-18T06:23:17Z
managedclusters.cluster.open-cluster-management.io             2021-11-18T06:23:17Z
managedclustersetbindings.cluster.open-cluster-management.io   2021-11-18T06:23:18Z
managedclustersets.cluster.open-cluster-management.io          2021-11-18T06:23:17Z
manifestworks.work.open-cluster-management.io                  2021-11-18T06:23:17Z
placementdecisions.cluster.open-cluster-management.io          2021-11-18T06:23:19Z
placements.cluster.open-cluster-management.io                  2021-11-18T06:23:18Z
llhu@huleis-mbp clusteradm % bin/clusteradm clean 
2021/11/18 14:23:40 Wait to delete cluster manager resource: cluster-manager.
2021/11/18 14:24:45 Recource cluster-manager have been deleted firstly.
The multicluster hub control plane has been clean up successfully!
llhu@huleis-mbp clusteradm % kubectl get crd 
No resources found
llhu@huleis-mbp clusteradm % kubectl get ns 
NAME                 STATUS   AGE
default              Active   11m
kube-node-lease      Active   11m
kube-public          Active   11m
kube-system          Active   11m
local-path-storage   Active   11m
```

File: pkg/cmd/clean/exec.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	if !o.ClusteradmFlags.DryRun {
+		b := retry.DefaultBackoff
+		b.Duration = 1 * time.Minute
+
+		err = WaitResourceToBeDelete(context.Background(), clusterManagerClient, o.clusterManageName, b)
+		if err != nil {
+			log.Println("Recource cluster-manager have been deleted firstly.")
```
Comment (qiujian16) [inline pkg/cmd/clean/exec.go:78]: should fatal here?

File: pkg/cmd/clean/exec.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	clusterManagerClient, err := clustermanagerclient.NewForConfig(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = clusterManagerClient.OperatorV1().ClusterManagers().Delete(context.Background(), o.clusterManageName, metav1.DeleteOptions{})
+	if err != nil {
+		log.Fatal(err)
```
Comment (qiujian16) [inline pkg/cmd/clean/exec.go:70]: you might consider IsNotFoundErr here, if it is not found. It should proceed instead of fatal

**Comment by xauthulei** [comment]
@qiujian16 , Updated it as your latest comments, Could you help review this. Thanks

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

thanks!

**Pull request merged**